db/compaction_picker.cc

```diff
@@ -315,13 +315,29 @@ Compaction* CompactionPicker::CompactFiles(
   // shouldn't have been released since.
   assert(!FilesRangeOverlapWithCompaction(input_files, output_level));
 
-  auto c =
-      new Compaction(vstorage, ioptions_, mutable_cf_options, input_files,
-                     output_level, compact_options.output_file_size_limit,
-                     mutable_cf_options.max_compaction_bytes, output_path_id,
-                     compact_options.compression, ioptions_.compression_opts,
-                     compact_options.max_subcompactions,
-                     /* grandparents */ {}, true);
+  CompressionType compression_type;
+  if (compact_options.compression == kDisableCompressionOption) {
+    int base_level;
+    if (ioptions_.compaction_style == kCompactionStyleLevel) {
+      base_level = vstorage->base_level();
+    } else {
```

Comment:
The comment for `VersionStorageInfo::base_level` indicates it's only populated when level compaction is used. Otherwise it is -1, which messes up the index computation into the `compression_per_level` array.

I verified this code works with all compaction styles (except FIFO for which it is irrelevant since that only supports L0->L0).

The source of truth is `compact_options`. If `compact_options` contains sentinel value, we fall back to `GetCompressionType`.